### ISSUE TITLE
ASSERT core/dynamo.c:1139 on running sol on Ubuntu

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#1216]
_From [zhao...@google.com](https://code.google.com/u/106321947286816917100/) on July 18, 2013 11:42:26_

What steps will reproduce the problem? 1. in Ubuntu 13.04: ./bin64/drrun -debug -c ./api/samples/bin/libbbcount.so -- sol
2. close the windows when it shows up What is the expected output? What do you see instead? stop the app successfully, but I saw assert failure

<Stopping application /usr/games/sol (14617)>

^T<CURIOSITY : loop_count \< max_loops in file /home/dynamorio/Workspace/DynamoRIO/dynamorio/core/synch.c line 1361
version 4.0.2186, custom build
-client_lib '/home/dynamorio/Workspace/DynamoRIO/builds/build_x64_dbg/./api/samples/bin/libbbcount.so;0;' -code_api -stack_size 56K -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct 
0x0000000049a66680 0x00000000711fef65
0x0000000049a66750 0x000000007108aefa
0x0000000049a66780 0x000000007108b02c
0x0000000049a667b0 0x000000007108b1d6
0x0000000049a667e0 0x0000000071298564
0x0000000049a667f8 0x00000000499d5e40
/home/dynamorio/Workspace/DynamoRIO/builds/build_x64_dbg/./api/samples/bin/libbbcount.so=0x0000000072000000
/lib/x86_64-linux-gnu/libc.so.6=0x00007fdb30364000
/lib64/ld-linux-x86-64.so.2=0x00007fdb3013f000>

<Application /usr/games/sol (14617).  Internal Error Internal DynamoRIO Error: /home/dynamorio/Workspace/DynamoRIO/dynamorio/core/dynamo.c:1139 ok
(Error occurred @164694 frags)
version 4.0.2186, custom build
-client_lib '/home/dynamorio/Workspace/DynamoRIO/builds/build_x64_dbg/./api/samples/bin/libbbcount.so;0;' -code_api -stack_size 56K -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct 
0x0000000049a665f0 0x0000000071111db9
0x0000000049a66750 0x000000007108af31
0x0000000049a66780 0x000000007108b02c
0x0000000049a667b0 0x000000007108b1d6
0x0000000049a667e0 0x0000000071298564
0x0000000049a667f8 0x00000000499d5e40
/home/dynamorio/Workspace/DynamoRIO/builds/build_x64_dbg/./api/samples/bin/libbbcount.so=0x0000000072000000
/lib/x86_64-linux-gnu/libc.so.6=0x00007fdb30364000
/lib64/ld-linux-x86-64.so.2=0x00007fdb3013f000>

It runs fine if without client.

_Original issue: http://code.google.com/p/dynamorio/issues/detail?id=1216_

**Comment by derekbruening** [comment]
A related issue here is our own signals interrupting auto-restart syscalls at attach time, where our #1145 auto-restart emulation does not come into play as we're interrupting native code.

A simple proposal here is to set SA_RESTART on our own handler.  This will cause the kernel to adjust the sigcontext to restore the original xax and shift the PC back by 2.  This should solve the attach-time problem, and should also help with in-cache and gencode cases.  There should be no downside, assuming DR can handle "before do_syscall": if that's not considered a safe point and we assume that thread will make progress we could have a problem.

I think the reason #1145 didn't use SA_RESTART was a mis-understanding of when the kernel adjusts the PC: if the kernel adjusts on sigreturn then it messes with cases where DR changes the PC.  But the kernel adjust prior to setting up the frame.

**Comment by derekbruening** [comment]
Blindly putting in SA_RESTART hits a hang on eintr-noinline test.

It records the signal, and returns -- but that return restarts the
syscall, which blocks, and we then never deliver the signal.

-ignore_syscalls doesn't hang b/c:

```
record_pending_signal(23) from cache pc 0x000000005535b044
        delaying until exit F2033
signal interrupted pre/post syscall itself so delivering now
```

So this is #1206 "must deliver signal received in do_syscall immediately"?
We canceled that b/c w/ the actual interruption we didn't need it.
Xref complications of post-syscall handling if we deliver a signal now (for
inlined doesn't matter b/c there are no pre or post handlers).

This goes back to another problem: we can't distinguish the case of not yet
having executed the syscall yet, from executed it but got interrupted --
though get_at_syscall() should distinguish them, right?

So the proposal is:
+ Have SA_RESTART
+ For interrupting a syscall in DR or client C code, no further work
  needed: just enjoy the lack of an EINTR loop.
+ For interrupting an inlined syscall in the cache: deliver the signal
  immediately as we do now, just as though the signal arrived just before
  the syscall executed.
+ For interrupting do_syscall: first check get_at_syscall() to confirm we
  executed the pre-syscall part.  If not, handle normally.  If so, we
  either have to go back to dispatch to deliver the signal, in which case
  we may as well undo the pc-2 and xax restore and use our auto-restart
  emulation code, or we deliver the signal immediately and have problems w/
  cross-syscall state (just like "CLS" on Windows).
  Having to keep emulating seems backward but it does seem like the
  simplest solution?  Or should we try to call handle_post_system_call()
  and then deliver the signal immediately?

Re: is it un-transparent to have EINTR returned for an auto-restart
syscall: since no app code is run and it's just DR and client-visible state
it seeems ok.  I guess a client counting syscall returns would have it
wrong.

**Comment by derekbruening** [comment]
This issue is really about the post-syscall translation point, which was solved by adding nops in 67e80387665c79f7df2984952c2cc168a19e814c.  I've split out the restart issues with native code into #2659.
